### ISSUE TITLE
Allow lazy random access to files in Storage on S3

### ISSUE THREAD

**Issue by brianhelba** [ResonantGeoData/ResonantGeoData#237]
When a Django `FieldFile` (the object referenced by a `FileField`) is opened, both in development with MinIO and in production with AWS S3, the `Storage` layers download a whole file write it to a `SpooledTemporaryFile` (which is in-memory for small content and on-disk for larger content)
* `S3Boto3Storage`: https://github.com/jschneier/django-storages/blob/770332b598712da27ecdba75c9e202ad6a1a8722/storages/backends/s3boto3.py#L132
* `MinioStorage`: https://github.com/py-pa/django-minio-storage/blob/8ce77cd8eae0d86fa3200e74cde64144c0338f37/minio_storage/files.py#L130

However, many operations need random access to only a small part of the file's content.

If all that's necessary is a Python file-like object (which is the same interface that a Django `FieldFile` naturally provides), we could look into developing a new `Storage` wrapper which lazily loads the content of the file-like object, internally using HTTP Range requests to only download the required parts just-in-time to be read from. This would likely also include some caching of the parts once they are retrieved.

If external tools (like GDAL) need access via a filesystem path (i.e. what's provided by `field_file_to_local_path`), then FUSE may be the most appropriate abstraction. However, FUSE could not be run on Heroku and might generally be more difficult to install for new developers.

In either case (Python file-like object or FUSE), the underlying logic for how to fetch and cache portions of file content may be the same.

**Comment by manthey** [comment]
If we are only accessing files from GDAL, we can use the `/vsis3` virtual file mode of GDAL.  I'm not sure that this can work if there is more than one S3 bucket involved as GDAL uses environment variables to determine some access.  This would set the environment variables `AWS_SECRET_ACCESS_KEY`, `AWS_ACCESS_KEY_ID`, `AWS_REGION`, `AWS_DEFAULT_REGION`, `AWS_S3_ENDPOINT` and the either spawn GDAL with that environment or open a file via the python wrapper.  The file name is `/vsis3/<bucket and path in s3>`.

**Comment by banesullivan** [comment]
Unfortunately, I do not think launching an environment with the correct S3 credentials for GDAL would suffice. In #249, I added *general support for files hosted at any URL*. While down the road we can implement better handling of `s3://` URLs and manage individual user S3 credentials, I think we want to broadly support lazy access to files at any URL.

At present, the following code pulls the entire file down locally during our ETL routines:

https://github.com/ResonantGeoData/ResonantGeoData/blob/1180d27ed038eb8add2d3a9b9e7b3bdd650db67a/rgd/utility.py#L163-L172

This is quite expensive when GDAL (or another lib) just needs to read a little bit of the file to retrieve the metadata (also I am aware of other problems with this such as caching which I will fix in a follow-up PR).

I would like this path generator to leverage FUSE to mount any file at a given URL to a path on disk and handle any read-only operations for accessing that file (regardless if that URL is for S3 or not). Also, this would enable us to handle other ETL routines not using GDAL for remote files.

**Comment by banesullivan** [comment]
FYI: I stumbled upon this module which allows lazy access to files at any URL (and it has support for S3): https://github.com/higlass/simple-httpfs

I managed to set this up and read a large image from remote storage with minimal effort:

```bash
python -m simple_httpfs /Users/bane/Desktop/landsat/foo/https
```

```py
from osgeo import gdal
import os

# Large-ish image URL:
url = 'https://s3-us-west-2.amazonaws.com/landsat-pds/c1/L8/033/038/LC08_L1TP_033038_20200101_20200101_01_RT/LC08_L1TP_033038_20200101_20200101_01_RT_B4.TIF'
path = url.replace('https://', 'foo/https/') + '..'

# This was timed
s = gdal.Open(path)
ds.RasterCount
```

```
145 µs ± 4.65 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

```py
>>> import rasterio as rio
>>> src = rio.open(path)
>>> src.meta
{'driver': 'GTiff',
 'dtype': 'uint16',
 'nodata': None,
 'width': 7751,
 'height': 7881,
 'count': 1,
 'crs': CRS.from_epsg(32613),
 'transform': Affine(30.0, 0.0, 220785.0,
        0.0, -30.0, 3630915.0)}
```

```
1.55 ms ± 51.4 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

------

Considering we do not currently have Heroku workers enabled and we are using a local worker launched (by me) via docker-compose, I could set up FUSE in our worker image and change `url_file_to_local_path` to leverage this if mounted or fall back to the original, expensive code.

**Comment by banesullivan** [comment]
This would also work to circumvent the usage of `field_file_to_local_path` on the `S3FieldFile` as we can just point to the file fields URL

**Comment by manthey** [comment]
Interesting that simple-httpfs explicitly reads 256 kb at a time from http when possible, caching up to 1 Gb of such chunks in a temp directory.  In the girder-3 mount, we rely on the OS doing the appropriate buffered reads.  By default, ubuntu reads 128 kb blocks.

The limitation of this library will probably be when file names are important and expected to be relative to one another (for instance, if you have a vrt file that references multiple raster files, or, for non-geospatial mirax files).

**Comment by banesullivan** [comment]
> The limitation of this library will probably be when file names are important and expected to be relative to one another (for instance, if you have a vrt file that references multiple raster files, or, for non-geospatial mirax files).

This indeed will be a limitation of this approach but IMO worth the cost to have lazy access for all other file types.

FWIW, if the files are hosted in a directory-like manner and the URLs reflect that, then we could still do relative paths (I'm actually doing that for the Landsat 8 imagery on S3 where the URLs mimic a directory structure).

I am going to proceed with a first pass at implementing this to boost data ingestion times and we can always refine down the road.

**Comment by mcovalt** [comment]
I hate to say it, but I vote against FUSE. I do agree FUSE is the most generic approach. It would be nice to have GDAL, Rasterio, Python, and any other library all speak the same language re: remote file access, but it could be big ask. We can't code our way around a sysadmin saying "no"! Here are the requirements for FUSE:
1. FUSE kernel module loaded
2. `CAP_SYS_ADMIN` privs for the Docker container

FUSE in Docker kinda breaks the seccomp jail and portability of Docker. Another option is [ParrotFS](http://ccl.cse.nd.edu/software/parrot/). It uses [`ptrace`](https://man7.org/linux/man-pages/man2/ptrace.2.html) to intercept and redirect syscalls like [`read`](https://man7.org/linux/man-pages/man2/read.2.html) etc. to a userspace implementation. Only a half joke... I think it's great! Of course it needs `CAP_SYS_PTRACE` which is also a big ask for a sysadmin.

I'm using [`/vsis3/`](https://gdal.org/user/virtual_file_systems.html#vsis3-aws-s3-files) and [`/vsicurl/`](https://gdal.org/user/virtual_file_systems.html#vsicurl-http-https-ftp-files-random-access) in #276. You can see it's implementation in 8b4d2fc6f162447c128b1a4fdb17e8353853dd09. Two big requirements to this approach:
1. **it only works for GDAL and Rasterio access**: e.g. [`osgeo.gdal.listdir`](https://gdal.org/python/osgeo.gdal-module.html#listdir), not [`os.listdir`](https://docs.python.org/3/library/os.html#os.listdir). annoying.
2. **external files are not writable**: I think this is ok?

The [`/vsicurl/`](https://gdal.org/user/virtual_file_systems.html#vsicurl-http-https-ftp-files-random-access) interface is really flexible. We can easily set headers per request for read-only access to external S3 buckets. Another neat feature: it'll follow redirects to presigned S3 urls. I'm not sure how the client is expecting to protect their resources, but I think [`/vsicurl/`](https://gdal.org/user/virtual_file_systems.html#vsicurl-http-https-ftp-files-random-access) can handle any variant as long as it's over http(s) or ftp.

Along the same thought, I think [`/vsizip/`](https://gdal.org/user/virtual_file_systems.html#vsizip-zip-archives) and [`/vsigzip/`](https://gdal.org/user/virtual_file_systems.html#vsigzip-gzipped-file) will work for the cases where we download→extract→upload files. We can use [all the above together via chaining](https://gdal.org/user/virtual_file_systems.html#chaining).

**Comment by manthey** [comment]
The real virtue of fuse (over something like ParrotFS) is if you write enough code to expose file paths in some determinable way then you can easily support file formats that use multiple files.  For instance, if we have a TIF file with an OVR file, on a local file system, opening the TIF file with GDAL will also access the OVR file and be able to potentially read appropriate lower resolution versions of an image from that.  With the httpfs implementation OR using vsis3, GDAL doesn't have enough information to find the OVR file.  Similarly, with ParrotFS, it wouldn't be found.  However, with the fuse setup we did in Girder 3, if the two files were in the same item, GDAL would read them properly.

I think the ptrace technique could be better as it doesn't require any change to the parent system (since there is no need for libfuse / dokany) -- but it would require at least as much work as we'd have to do for fuse to get support for adjacent files.  Since it is much less common than fuse, I suspect it will be harder to maintain, too.  And, if we ever need a read-write file system, I know how to do that in fuse but would have to learn how to do that the ptrace way.

**Comment by banesullivan** [comment]
Notes from team discussion on this:

We are going to stick with using FUSE in the worker instance in an optional manner and `vsicurl` for server tiles in the app instance

#### Worker Instance

FUSE will be implemented such that when fetching a path for a file field:
1. we will check if FUSE is available and return a FUSE mounted path to the file resource regardless of whether it is in our S3 storage, minio, or a remote location specified by an HTTP URL.
2. if FUSE is unavailable, we will have an option to fall back to a `vsicurl` path
3. otherwise we will utilize `field_file_to_local_path` to expensively download the entire file locally for files in S3FileField and `url_file_to_local_path` for URL file resources (implemented in #257)

When we deploy our worker instance on EC2, it will be set up with FUSE but we need to have the option to run the worker in non-privileged environments as well.

#### App Instance

Utilizing `vsicurl` in the app instance will work for now because we are only using GDAL as the tile source. By avoiding FUSE and sticking with `vsicurl`, we can continue to run the app in a non-privileged manner on Heroku and still have the advantages of lazy/random file access to quickly serve tiles. This is implemented in #276. If down the road we switch to needing other tile sources (e.g. mapnik) then we will have to leverage FUSE and switch to a custom deployment option off of Heroku (unlikely anytime soon).

**Comment by banesullivan** [comment]
Follow-up concern: We are still computing checksums of all files. Should we stop doing this to avoid downloading entire large files?
